lib/pos_mempool.go

```diff
@@ -921,82 +910,73 @@ func (mp *PosMempool) validateTransactions() error {
 	}
 
-	// Convert the mempool transactions to the MsgDeSoTxn format, which we can use for connecting to the validation view.
-	var txns []*MsgDeSoTxn
-	var txHashes []*BlockHash
-	for _, txn := range mempoolTxns {
-		txns = append(txns, txn.Tx)
-		txHashes = append(txHashes, txn.Hash)
-	}
-	// Copy the validation view to avoid modifying the readOnlyLatestBlockView.
-	copyValidationView := validationView.CopyUtxoView()
-	// Connect the transactions to the validation view. We use the latest block height + 1 as the block height to connect
-	// the transactions. This is because the mempool contains transactions that we use for producing the next block.
-	_, _, _, _, errorsFound, err := copyValidationView.ConnectTransactionsFailSafeWithLimit(
-		txns, txHashes, uint32(mp.latestBlockHeight)+1,
-		time.Now().UnixNano(), true, false, true, mp.maxValidationViewConnects)
-	if err != nil {
-		return errors.Wrapf(err, "PosMempool.validateTransactions: Problem connecting transactions")
-	}
-
-	// We iterate through the successFlags and update the validated status of the transactions in the mempool.
-	var txnsToRemove []*MempoolTx
-	for ii, errFound := range errorsFound {
-		if ii >= len(mempoolTxns) {
+	// Iterate through all the transactions in the mempool and connect them to copies of the validation view.
+	for ii, txn := range mempoolTxns {
+		// Break out if we've attempted to connect the maximum number of txns to the view
+		if uint64(ii) >= mp.maxValidationViewConnects {
 			break
 		}
-		// If the transaction successfully connected to the validation view, we update the validated status of the
-		// transaction in the mempool. If the transaction failed to connect to the validation view, we add it to the
-		// txnsToRemove list. Note that we don't need to hold a lock while updating the validated status of the
-		// transactions in the mempool, since the updateTransactionValidatedStatus already holds the lock.
-		if errFound == nil {
-			mp.updateTransactionValidatedStatus(mempoolTxns[ii].Hash, true)
-		} else {
-			txnsToRemove = append(txnsToRemove, mempoolTxns[ii])
-			// Add an error for the txn to our cache so we can return it to the user if they
-			// ask for it later.
-			mp.recentRejectedTxnCache.Add(*mempoolTxns[ii].Hash, errFound)
-		}
-	}
 
-	// Now remove all transactions from the txnsToRemove list from the main mempool.
-	mp.Lock()
-	for _, txn := range txnsToRemove {
-		if err := mp.removeTransactionNoLock(txn, true); err != nil {
-			glog.Errorf("PosMempool.validateTransactions: Problem removing transaction with hash (%v): %v", txn.Hash, err)
+		// Connect the transaction to a copy of the validation view. We can skip signatures on the transaction
+		// connect if the transaction has already previously been validated and been found to have a valid
+		// signature. This optimizes the connect by not repeating signature verification on a transaction
+		// more than once.
+		//
+		// TODO: Nina to confirm the signature validation assumption.
```

Comment:
@lazynina - I *think* this is safe. Provided the transaction has been validated once, it MUST have a valid signature

Reply:
The only problem is transactions signed by derived key. It's possible that a derived key has been de-authorized between runs of the transaction validation routine. But then the transaction would also fail validation because it wouldn't have sufficient spending limits. I need to think on this more, but I think it's okay to do.

Reply:
Sounds good @lazynina. Based on the cpu profiling, it looks like biggest cpu overhead in the mempool is repeat signature verification.

And it makes sense why that would be. This validation goroutine runs every 10ms and validates each txn multiple times in between the 1.5sec block time.

Reply:
This seems fine. If the derived key was de-authorized, it'll get caught by the spending limits check and marked invalid. There should never be a need to check signatures a second time.